Merge ckeditor improvements - Meeds-io/MIPs#71

hakermi opened the pull request:
Merge ckeditor improvements

hakermi committed:
feat: Notes editor is not wysiwyg - EXO-65725 - Meeds-io/MIPs#70 (#713)

Notes editor is not wysiwyg

feat: Notes editor is not wysiwyg - EXO-65725 - Meeds-io/MIPs#71 (#721)

Notes editor is not wysiwyg

fix: Adjust blockquote wysiwyg styles - EXO-65725 - Meeds-io/MIPs#71 (#732)

Adjust blockquote wysiwyg styles

fix: Adjust blockquote of list of items styles - EXO-65725 - Meeds-io/MIPs#71 (#737)

Adjust blockquote of list of items styles

fix: blockquotes in notes view mode are not wysiwyg - EXO-65725 - Meeds-io/MIPs#71 (#744)

Prior to this change, blockquotes in notes view mode are not wysiwyg in
term of font style. The PR makes sure to have the same normal font style
in both view and edit mode

hakermi committed:
feat: edito page and notes overview page size are not consistent - EXO-65734 - Meeds-io/MIPs#70 (#715)

edito page and notes overview page size are not consistent

hakermi committed:
feat: Notes editor toolbar options improvments - EXO-65695 - Meeds-io/MIPs#71 (#725)

Notes editor toolbar options improvments

fix: Adjust editor toolbar options order - EXO-65695 - Meeds-io/MIPs#71 (#734)

Adjust editor toolbar options order

hakermi committed:
feat: Add more colors option to text color button of editor toolbar - EXO-65704 - Meeds-io/MIPs#71 (#727)

Add more colors option to text color button of editor toolbar

hakermi committed:
feat: Allow only self and blank selectable options in cke link target dialog - EXO-65705 - Meeds-io/MIPs#71 (#728)

Allow only self and blank selectable options in cke link target dialog

fix: Prevent modifying links target attribute set by the link plugin - EXO-65705 - Meeds-io/MIPs#71 (#731)

Prevent modifying links target attribute set by the link plugin

hakermi committed:
feat: Enable search by tags for notes - EXO-65706 - Meeds-io/MIPs#71 (#736)

Enable search by tags for notes

hakermi committed:
feat: Notes editor cke table option - EXO-65716 - Meeds-io/MIPs#71 (#742)

Add Notes editor cke table option

fix: fix cke table alignment and header style - EXO-65716 - Meeds-io/MIPs#71 (#745)

Prior to this change, table alignment is not wysiwyg and header font
size is less than normal column.
This PR adjusts the alignment in header style issue.

fix: fix cke table option styles - EXO-65716 - Meeds-io/MIPs#71 (#750)

Prior to this change, inserted tables are not wysiwy, This PR adds some
adjustements for table styles in order to make it consistent in view and
edit mode

hakermi committed:
feat: Indent option is available only in lists - EXO-66446 - Meeds-io/MIPs#71 (#747)

Prior to this change, indent option is available only in lists due to
missing indent plugins.
This PR adds the needed indent extra plugins to make it available on all
possible items.

hakermi committed:
fix: fix inserted image wysiwyg issue in edit mode - EXO-66452 - Meeds-io/MIPs#71 (#749)

Prior to this change, inserted images in edit mode are not consistent
with images in view mode due to the added styles of spacement of the
widget container, resize and drag buttons.
This PR adds some adjustments inside the image widget container to make
it consistent with the images in the edit mode

hakermi committed:
feat: Add cke accessibility plugin - EXO-66514 - Meeds-io/MIPs#71 (#753)

Add cke accessibility plugin

hakermi committed:
feat: codenippets plugin - EXO-65719 - Meeds-io/MIPs#71 (#756)

Add codesnippets plugin

hakermi committed:
feat: Allow open inserted note navigations in a new tab - EXO-66034 - Meeds-io/MIPs#71

Allow open inserted note navigation in a new tab

hakermi committed:
feat: set zero margin top of first title comes in page content - EXO-66634 - Meeds-io/MIPs#71 (#761)

set zero margin top of first title comes in page content

hakermi committed:
feat: Add analytics on added tags metadata in notes content - EXO-66540 - Meeds-io/MIPs#71 (#762)

Add analytics on added tags metadata in notes content

fix: fix test failure - EXO-66540 (#769)

fix test failure

hakermi committed:
feat: Add assets version to editor content css file - EXO-65716 - Meeds-io/MIPs#71

Add assets version to editor content css file

hakermi committed:
fix: Adjust wysiwyg issue in content area between view and edit mode -  EXO-66650 -  Meeds-io/MIPs#71 (#766)

Prior to this change, there is a difference of 2px in terms of width
inside the content area between the edit and view mode which cause a
wysiwyg.
This PR adjusts the margin set in the editor content area to have the
same as the view mode.

hakermi committed:
fix: wysiwy issue in table border property EXO-66692 - Meeds-io/MIPs#71 (#771)

Prior to this change, changing the border property is not reflected on
view mode.
This PR ensures to reflect the change of border property from plugin
dialog on view mode

hakermi committed:
fix: set zero margin top of first title comes in page content - EXO-66864 - Meeds-io/MIPs#71 (#801)

This PR removes the !important flag that is used to override the
margin-top of first titles in the page content to have a "0" margin top.

hakermi committed:
fix: wysiwyg issue in tabel when width property is not set - EXO-66882 - Meeds-io/MIPs#71 (#805)

wysiwyg issue in tabel when width property is not set

hakermi committed:
feat: Centrelize ckeditor js config and styles - EXO-66890 - Meeds-io/MIPs#71 (#838)

Centrelize ckeditor js config and styles

hakermi committed:
feat: Enhance embedding media display in mobile - EXO-65717 - Meeds-io/MIPs#71  (#840)

Enhance embedding videos display in mobile

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/jpa/search/WikiElasticSearchServiceConnector.java:
```diff
@@ -70,6 +70,15 @@ public class WikiElasticSearchServiceConnector extends ElasticSearchServiceConne
 
   private String                     searchQueryFilePath;
 
+  public static final String           SEARCH_QUERY_TERM            = ",\"must\":{" +
+          "  \"query_string\":{" +
+          "    \"fields\": [\"name\",\"title\",\"content\",\"comment\",\"attachment.content\"]," +
+          "    \"default_operator\": \"AND\"," +
+          "    \"query\": \"@term@\"" +
+          "  }" +
+          "}";
```
You can use `triple double quote sequence` to simplify reading and maintenance:

```suggestion
  public static final String           SEARCH_QUERY_TERM            = """
         ,"must":{
            "query_string":{
              "fields": ["name","title","content","comment","attachment.content"],
              "default_operator": "AND",
              "query": "@term@"
            }
          }
  """;
```

hakermi replied:
done

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/jpa/search/WikiElasticSearchServiceConnector.java:
```diff
@@ -111,33 +120,67 @@ protected String getSourceFields() {
+      return "";
+    }
+    List<String> tagsQueryParts = values.stream()
+            .map(value -> "{\"term\": {\n" +
```
Same, you can use `triple double quote sequence` to simplify reading and maintenance

hakermi replied:
done

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/listener/AnalyticsAddTagListener.java:
```diff
@@ -0,0 +1,49 @@
+package org.exoplatform.wiki.listener;
```
Missing license Header
(To apply on all classes please to not repeat the comment multiple times)

hakermi replied:
done

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/listener/AnalyticsAddTagListener.java:
```diff
@@ -0,0 +1,49 @@
+package org.exoplatform.wiki.listener;
```
For new classes in Meeds, please use package name `io.meeds` instead of `org.exoplatform`
(To apply on all classes please to not repeat the comment multiple times)

```suggestion
package io.meeds.wiki.listener;
```

hakermi replied:
done

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -132,6 +132,7 @@ public class NoteServiceImpl implements NoteService {
 
   private final HTMLUploadImageProcessor                  htmlUploadImageProcessor;
 
+
```
useless line break
```suggestion
```

hakermi replied:
done

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -242,13 +242,13 @@ public Page createNote(Wiki noteBook, String parentNoteName, Page note, Identity
     }
   }
-
+  
```
useless space

```suggestion
```

hakermi replied:
done

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -275,6 +275,7 @@ public Page updateNote(Page note, PageUpdateType type, Identity userIdentity) th
       }
     }
     Page updatedPage = updateNote(note, type);
+    Utils.broadcast(listenerService, "note.updated", userIdentity.getUserId(), updatedPage);
```
Shouldn't we trigger/centralize this event in method `public Page updateNote(Page note, PageUpdateType type)`, WDYT ?

hakermi replied:
I used the updatePage function that is called in the rest layer in all update actions while the above method is used only in the service layer.
And the important reason is that i need the `userIdentity` which is not available in this method.

boubaker replied:
Yep, this is not the good pattern to use for event triggering .... When the update id made by a batch update (auto import when server starts up by example), the listener should be triggered as well if if no user had auctioned the update.

hakermi replied:
Ok but without the userId the listener is useless
https://github.com/Meeds-io/notes/pull/843/files#diff-3198ed87ff299273c36b60c054d7cfdef2b2eada4a3f90bd54d51fe498237cb7R61-R62

boubaker replied:
Is it really?

```java
    Page note = event.getData();
    String username = event.getSource();
    long creatorId = getPosterId(username);
    long audienceId = getStreamOwnerId(note.getWikiOwner(), username);

    Set<TagName> tagNames = tagService.detectTagNames(note.getContent());
    tagService.saveTags(new TagObject(NOTES_METADATA_OBJECT_TYPE, note.getId(), note.getParentPageId()),
                        tagNames,
                        audienceId,
                        creatorId);
    indexingService.reindex(WikiPageIndexingServiceConnector.TYPE, note.getId());
```

The listener should:
* update Tags
* reindex the page

**independently** whether the action is made **by a user** or **by a batch process**, those updates has to be made.

Shouldn't we delete the `indexingService.reindex` since the note page is automatically reindexed when a metadata item is modified (triggered using listener `org.exoplatform.wiki.service.listener.MetadataItemModified`) ?

hakermi replied:
I didn't understand,
```
 tagService.saveTags(new TagObject(NOTES_METADATA_OBJECT_TYPE, note.getId(), note.getParentPageId()),
                        tagNames,
                        audienceId,
                        creatorId
```); 

In the createTags isn't the userIdentityId mandatory ? 
how i can get it without passing it as parameter in the event?

boubaker replied:
you can use the last updater or the creator, it's not really important information that will be used after saving the associated Tags to a note, the most important information is the `audienceId`

hakermi replied:
There is no creator or updater property in the note page and the `page.getAuthor()` is always null because there is no author property in `pageEntity`

boubaker replied:
> There is no creator or updater property in the note page and the `page.getAuthor()` is always null because there is no author property in `pageEntity`

I let you find a solution, for my part the flow and layers are important to preserve coherence and information consistency.

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/utils/Utils.java:
```diff
@@ -684,4 +692,11 @@ public static String html2text(String html) {
+    try {
+      listenerService.broadcast(eventName, source, data);
+    } catch (Exception e) {
+      LOG.error("Error while broadcasting event: {}", eventName, e);
```
Use warn log level when the operation will not break the rest of computation. This will be used just for logging and troubleshooting.

```suggestion
      LOG.warn("Error while broadcasting event: {}. Wheras, the operation will continue to proceed", eventName, e);
```

hakermi replied:
done

boubaker commented on notes-webapp/src/main/webapp/WEB-INF/conf/wiki/metadata-configuration.xml:
```diff
@@ -0,0 +1,42 @@
+<!--
+    Copyright (C) 2023 eXo Platform SAS.
```
Wrong license Header

```suggestion
```

hakermi replied:
fixed

boubaker commented on notes-webapp/src/main/webapp/vue-app/notes/components/NotesOverview.vue:
```diff
@@ -199,8 +199,8 @@
-              <v-list-item-title @click="openNoteChild(item)" class="body-2 clickable primary--text">
-                <span>{{ item.name }}</span>
+              <v-list-item-title class="body-2 content-link clickable primary--text">
+                <a :href="getNoteLink(item.noteId)"><span>{{ item.name }}</span></a>
```
For a common UX with other parts of the product, can you use please `<v-list-item :href` instead of `<v-list-item-title....<a :href`

```suggestion
              <v-list-item :href="getNoteLink(item.noteId)" class="body-2 primary--text">
                {{ item.name }}
              </v-list-item>
```

Besides, to not use a method here and use a computed instead, can you add a dedicated component for Tree leaf.

hakermi replied:
done and to check

boubaker commented on notes-webapp/src/main/webapp/vue-app/notes/components/NotesOverview.vue:
```diff
@@ -809,13 +813,17 @@ export default {
       const contentChildren = docElement.getElementsByTagName('body')[0].children;
       const links = docElement.getElementsByTagName('a');
       const tables = docElement.getElementsByTagName('table');
+      docElement.querySelectorAll('pre > code').forEach((e) => codeHighlighter.highlightBlock(e));
```
Using a 149KB of new framework + its corresponding CSS seems not relevant. Please try to delete the usage of this framework if it's not a major functional requirement, at least in "**View Mode**".

```suggestion
```

hakermi replied:
> Using a 149KB of new framework + its corresponding CSS seems not relevant. Please try to delete the usage of this framework if it's not a major functional requirement, at least in "**View Mode**".

This library comes with ckeditor and used in the codesnippet plugin, so i created an amd library of it + a portal skin of the choose style of the code to ensure the WYSIWYG for codesnippet because there is no other way to do that in view mode. ckidetor.getData() doesn't return the exactly highlighted html code.
https://github.com/Meeds-io/commons/pull/551/commits/90092cc73ff14e478395b4f844cd14ed5c152ddf#diff-60986d51a51cfa41dd1f469bd45a68212c4512d265ab7dfb946a3127b6671862R5

boubaker replied:
ok for Style but no need to make the framework work in **View Mode**
The plugin is intended to generate an HTML snippet that has to be saved as is. And when doing so, no need to retrigger the plugin in View mode to have a good display of the injected HTML.

hakermi replied:
Ok but without the userId the listener is useless
 https://github.com/Meeds-io/notes/pull/843/files#diff-3198ed87ff299273c36b60c054d7cfdef2b2eada4a3f90bd54d51fe498237cb7R61-R62

hakermi replied:
OK but even if we keep the styles, in the view mode we will not have the same display of the code as in edit mode, as all highlighted parts are transformed into span tags with css classes, but these span tags are removed when call editor.getData()

mkrout replied:
And if we save the generated HTML, we will not be able to edit it again.

boubaker replied:
> And if we save the generated HTML, we will not be able to edit it again.

We have to not use `HTMLSanitizer` anymore when writing the information. (we don't have to loose the original information entered by the end user)
But we use `v-sanitized-html` to sanitize the html on rendering in order to preserve the original information.

hakermi replied:
The problem is not in the `HtmlSanitizer` it's in the ckeditor itself, in the `editor.getData()`

boubaker replied:
> And if we save the generated HTML, we will not be able to edit it again.

>The problem is not in the HtmlSanitizer it's in the ckeditor itself, in the editor.getData()

This is may be a cross cutting concern about CKEditor contents. We may need to preserve the HTML as displayed in CKEditor (`CKEDITOR.instances.**INSTANCE_NAME**.document.$.body.innerHTML`) to preserve the WYSIWYG effect, not only for notes. At the same time, we will need to preserve the response for `editor.getData()` to be able to edit the content with CKEditor plugins.

This would be a good alternative to [PHP backend processor](https://ckeditor.com/docs/ckeditor4/latest/features/codesnippetgeshi.html#back-end-configuration) of a version of this plugin.

We will talk more about this in the dedicated meeting.

hakermi replied:
I have already used something similar to preserve the embedded iframe of the media plugin
https://github.com/Meeds-io/notes/pull/843/files#diff-43927e45889ed109fe9f553ea38b02993f822a8fb1827fa9d26c660fa5cf3c79R813-R834
should i try to do the same for highlighter too ?

boubaker replied:
The idea is to minimize the **CPU & Memory consumption**, in browser side, when displaying Notes in Page. It should be as much static as possible to optimize the necessary ressources to display a page with multiple **SNVs** based on Notes.
Having to load a **heavy JS library** in addition to a computing to transform the HTML each time we display the page wouldn't be performant.

hakermi replied:
@boubaker PR updated
https://github.com/Meeds-io/notes/pull/843/commits/86836dc8b9e90bb08c614b8282ac7f50314ef4f2

hakermi committed:
fix: Fix feedback in ckeditor improvement - Meeds-io/MIPs#71

fix feedback on notesMetadadataListener and codehighliter plugin

Fix feedback on media embed and listeners invocation

fix feedback on styles and listeners invocation

boubaker commented on notes-service/src/main/java/io/meeds/notes/listener/NotesMetadataListener.java:
```diff
@@ -0,0 +1,90 @@
+    Page note = event.getData();
+    String username = event.getSource();
+    if (username == null) {
+      username = userACL.getSuperUser();
```
Can you use `note.getAuthor()` first if it's not null + ensure to populate the attribute when updating/creating it.
When author is null as well (like when importing a note from sources in server startup), in this special case, we will have to use `username = userACL.getSuperUser();`

hakermi replied:
note.getAuthor() is always null i said previously, because there is not author property in the pageEntity

boubaker replied:
yes, as agreed @mkrout @rdenarie in meeting, we should populate it with the last author name.

hakermi replied:
I already used a lastUpdater attribute and send it in the broadcast why i should use the author ?
`Utils.broadcast(listenerService, "note.posted", **note.getLastUpdater()**, createdPage);`

boubaker replied:
even if it's already covered at this time by the caller method, this method should be auto sufficient and as much coherent as possible by embedding inside it the full business logic, which is:
1. If event.getData() isn't null, then use it
2. Else if page.getAuthor isn't null, then use it
3. Else, when the page is new and was added by a batch method (like import pages process), use a default author which is the super admin

This way we are sure to have secured the business logic even when a future maintenance/feature is made on the caller method where the author isn't passed as event data object.

boubaker commented on notes-service/src/main/java/io/meeds/notes/listener/NotesMetadataListener.java:
```diff
@@ -0,0 +1,90 @@
+                        tagNames,
+                        audienceId,
+                        creatorId);
+    indexingService.reindex(WikiPageIndexingServiceConnector.TYPE, note.getId());
```
Please split this listener into two different listeners, each one will consider one Task to do at the same time. Using this combination of Tasks inside a single Listener will lead to not index the content when the Tags throws an error. This is a **best practice** the fact to split into multiple listeners and avoid combining them.

hakermi replied:
you want me to remove the `indexingService.reindex(...` right ? OK i will

hakermi replied:
I have removed the reindex because it happens in the `MetaDataModified` listener even if it doesn't make sense that when the `detectTags` fails(**task1**) we should `reindex`(**task2**).

boubaker replied:
> it doesn't make sense that when the detectTags fails(task1) we should reindex(task2).

Shouldn't we have the content of the updated note searchable when the Tags aren't correct ? Two different Use Cases that shouldn't block each other.

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -242,7 +242,7 @@ public Page createNote(Wiki noteBook, String parentNoteName, Page note, Identity
       // call listeners
       postAddPage(noteBook.getType(), noteBook.getOwner(), note.getName(), createdPage);
-
+      Utils.broadcast(listenerService, "note.posted", note.getLastUpdater(), createdPage);
```
We should broadcast listeners as well in `public Page createNote(Wiki noteBook, Page parentPage, Page note)` to ensure information consistency & coherence even in batch method.

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -297,15 +297,8 @@ public Page updateNote(Page note, PageUpdateType type) throws WikiException {
-    }
-
     Page updatedPage = getNoteById(note.getId());
+    Utils.broadcast(listenerService, "note.updated", note.getLastUpdater(), updatedPage);
```
We should broadcast listeners as well in `public Page updateNote(Page note)` to ensure information consistency & coherence even in batch method.

boubaker commented on notes-webapp/src/main/webapp/vue-app/notes/components/NoteContentTableItem.vue:
```diff
@@ -0,0 +1,23 @@
+  },
+  computed: {
+    noteLink() {
+      const baseUrl = window.location.href;
+      return this.note && `${baseUrl.substring(0, baseUrl.lastIndexOf('/') + 1)}${this.note.noteId}`;
```
We should use the `url` attribute returned in REST response. If the REST doesn't return it, we should fix the REST endpoint.

```suggestion
      return this.note?.url;
```

hakermi replied:
I have fixed the null url, but the whole idea about replacing with list-item has killed the ui of the table content.
beside a black color that is set in vuetify-all with !important flag in the list item there is also a default grey hover effect

boubaker replied:
> I have fixed the null url, but the whole idea about replacing with list-item has killed the ui of the table content.

Yes, that's the idea of using href on `v-list-item`, see it in multiple UIs, like Top bar menu:

![image](https://github.com/Meeds-io/notes/assets/594824/1040c46b-c6a3-4a20-b4c2-cf62d76a9df3)

This aspect is to validate with PMs, I let you discuss about this internally and decide which style to use.

But for the URL retrieving from REST endpoint instead of building it here, knowing that we already have this in REST while getting a single note, would be better for code coherence. In fact, getting a single `NotePage` of a list of `NotePage`s should retrieve the same set of information, except when we use a different `expand` parameter.

boubaker commented on notes-webapp/src/main/webapp/vue-app/notes/components/NotesOverview.vue:
```diff
@@ -809,13 +807,16 @@ export default {
       const contentChildren = docElement.getElementsByTagName('body')[0].children;
       const links = docElement.getElementsByTagName('a');
       const tables = docElement.getElementsByTagName('table');
+      docElement.querySelectorAll('oembed').forEach((oembed) => {
+        oembed.innerHTML = oembed.dataset.iframe;
+        delete oembed.dataset.iframe;
+      });
```
Sorry, I didn't get why we still need to change the saved HTML to make the display works. Can't we get a complete HTML from REST endpoint and display it as is ?
We will have a whole **CMS set of features** on top of this work. We should take the time to improve the performances of contents rendering now, to not have to make heavy and a substantial work for data migration just after few months. If we don't optimize the maximum of things now, we will probably never prioritize the performance improvements later when we will see that the page loading takes a lot of time when it's full of **SNVs**.

hakermi replied:
Same as codeSnippets plugin, the cke media plugin needs a processing as well, because when we insert a link using the embed media plugin, in edit mode  all is OK, but when call `editor.getData()` we have only "`<oembed>usedLink</oembed>`"

hakermi replied:
I have had to do something about this : https://github.com/Meeds-io/notes/pull/843/commits/a71cb13e8f8313ba82d6e6d17e232759bfdf8ba7#diff-43927e45889ed109fe9f553ea38b02993f822a8fb1827fa9d26c660fa5cf3c79R811-R852

`restoreEmbedded()` and `restoreunHighlited`() are needed to make the content compatible again with ckeditor in edit mode and recognized by the corresponding plugins

boubaker replied:
ok, no need to do more, we have to re-architecture the whole stuff to enhance this, by adding by example an additional field in DB for `Content To Edit` & keep the current one for `Content To View`. This would be the good integration to make with CKEditor advanced plugins that was added by current MIP... we will need to rework this but may be in the next MIP. Let's just keep it a simple implementation for now... Thanks.

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/model/Page.java:
```diff
@@ -98,6 +98,8 @@ public class Page {
   private Page                            parent;
 
   private Map<String, List<MetadataItem>> metadatas;
+  
+  private String                          lastUpdater;
```
As agreed in a meeting with @mkrout and @rdenarie we will reuse the `author` instead to have the last modifier, can we use it instead ?

boubaker commented on notes-webapp/src/main/webapp/vue-app/notes-editor/components/NotesEditorDashboard.vue:
```diff
@@ -853,9 +808,67 @@ export default {
+          const aspectRatio = width / height;
+          const minHeight = parseInt(this.oembedMinWidth) / aspectRatio;
+          const style = `
+          min-height: ${minHeight}px;
+          min-width: ${this.oembedMinWidth}px;
+          width: 100%;
+          margin-bottom: 10px;
+          aspect-ratio: ${aspectRatio};
+        `;
```
Formatting for code clearness

```suggestion
            min-height: ${minHeight}px;
            min-width: ${this.oembedMinWidth}px;
            width: 100%;
            margin-bottom: 10px;
            aspect-ratio: ${aspectRatio};
          `;
```

boubaker commented on notes-webapp/src/main/webapp/skin/less/notes/notes.less:
```diff
@@ -7,7 +7,8 @@
 @import "Sprites/Style.less";
 @import "UIWikiRichTextEditor/Stylesheet.less";
-
+@import "../../../ckeditor/skins/common/extendedRichContent.less";
+@import (less) "../../../ckeditor/plugins/codesnippet/lib/highlight/styles/monokai_sublime.css";
```
Not sure that this is the best place to define the reusable Less files, since we have platform-ui for it. I will change it anyway when starting to work on SNVs V2. Nothing to do more from your part, I 'll take it from here, but just to share the best practice for future implementations.

boubaker commented on notes-webapp/src/main/webapp/skin/less/notes/notes.less:
```diff
@@ -60,18 +61,20 @@
     background: #eeeeee;
 
+    .notes-content-wrapper {
+      min-height: calc(~"100vh - 210px");
```
there is new way introduced to manage Mobile ViewPort bug ([see it here](https://medium.com/@alekswebnet/fix-mobile-100vh-bug-in-one-line-of-css-dynamic-viewport-units-in-action-102231e2ed56))

```suggestion
      min-height: ~"calc(var(--100vh, 100vh) - 210px)";
```

boubaker commented on notes-webapp/src/main/webapp/skin/less/notes/notes.less:
```diff
@@ -60,18 +61,20 @@
+
     .notes-content {
-      height: calc(~"100vh - 120px");
+      height: calc(~"100vh - 100px");
```
same, even if we shouldn't use the viewport multiple times in the same page, but define it in the parent container and then use flexbox to adjust using `flex-grow-1`, `flex-grow-0`, `flex-column`...

```suggestion
      height: ~"calc(var(--100vh, 100vh) - 100px)";
```

boubaker commented on notes-webapp/src/main/webapp/skin/less/notes/notes.less:
```diff
@@ -60,18 +61,20 @@
       .notes-content-form {
-        background: white;
-        margin: 20px 14% 0 14% !important;
+        margin: 20px auto !important;
```
Can you avoid defining margins in CSS and move this into HTML using `mx-.my.ma.mt-...`

hakermi replied:
here i can't have exactly 20px with veutify helpers

boubaker commented on notes-webapp/src/main/webapp/skin/less/notes/notes.less:
```diff
@@ -341,6 +291,12 @@
+
+  .content-link {
+    a {
+      font-size: 16px !important;
```
We already have CSS helpers to define font sizes of elements, can you use it please instead.
Imagine we want to change the fonts sizes globally, we will have to change in one single place, plus this strategy avoids to have a huge CSS files to parse in browser side (CPU & Memory) and to download (Network performances)

hakermi replied:
which helper that allow to set exactly 16px as fontsize ?

boubaker replied:
`subtitle-1` defines in rem defines the style of subtitles used in product.

boubaker commented on notes-service/src/main/java/io/meeds/notes/listener/NotesMetadataListener.java:
```diff
@@ -0,0 +1,91 @@
+  public void onEvent(Event<String, Page> event) {
+    Page note = event.getData();
+    String username = event.getSource();
+    if (username == null && note.getAuthor() != null) {
+      username = note.getAuthor();
+    } else if (username == null) {
```
To simplify code clearness, thus to simplify code maintenance

```suggestion
    if (username == null) {
      username = note.getAuthor();
    }
    if (username == null) {
```

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -251,13 +250,16 @@ public Page createNote(Wiki noteBook, String parentNoteName, Page note, Identity
-
-    return dataStorage.createPage(noteBook, parentPage, note);
+    Page createdPage = dataStorage.createPage(noteBook, parentPage, note);
+    Utils.broadcast(listenerService, "note.posted", note.getAuthor(), createdPage);
```
We should trigger `postAddPage` as well

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -251,13 +250,16 @@ public Page createNote(Wiki noteBook, String parentNoteName, Page note, Identity
   public Page updateNote(Page note) throws WikiException {
-    return dataStorage.updatePage(note);
+    Page updatedPage = dataStorage.updatePage(note);
+    Utils.broadcast(listenerService, "note.updated", note.getAuthor(), updatedPage);
```
We should trigger `postUpdatePage` as well

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java:
```diff
@@ -284,6 +287,7 @@ public Page updateNote(Page note, PageUpdateType type, Identity userIdentity) th
     updatedPage.setAppName(note.getAppName());
     Map<String, List<MetadataItem>> metadata = retrieveMetadataItems(note.getId(), userIdentity.getUserId());
     updatedPage.setMetadatas(metadata);
+    postUpdatePage(updatedPage.getWikiType(), updatedPage.getWikiOwner(), updatedPage.getName(), updatedPage, type);
```
Already triggered in common method `updateNote(note)`, so we should avoid to trigger it twice

```suggestion
```

hakermi replied:
In the other method there is no type param which is needed in the GamificationListener 
https://github.com/Meeds-io/notes/blob/develop/notes-social-integration/src/main/java/org/exoplatform/wiki/integration/gamification/GamificationWikiListener.java#L99

boubaker replied:
In this case we centralize the real implementation of update in method `updateNote(Page note, PageUpdateType type)` (with **storage** and **listeners** call) and change the implementation of `updateNote(Page note)` this way :

```java
  public Page updateNote(Page note) throws WikiException {
    return updateNote(note, null);
  }
```

boubaker replied:
in fact, it would even be better to centralize all in the **full method** `Page updateNote(Page note, PageUpdateType type, Identity userIdentity)` and change the others this way:

```java
  public Page updateNote(Page note) throws WikiException {
    return updateNote(note, null);
  }

  public Page updateNote(Page note, PageUpdateType type) throws WikiException {
    return updateNote(note, type, null);
  }
```

boubaker commented on notes-service/src/main/java/org/exoplatform/wiki/service/rest/NotesRestService.java:
```diff
@@ -384,6 +384,7 @@ public Response createNote(@Parameter(description = "note object to be created",
       note.setSyntax(syntaxId);
       note.setName(TitleResolver.getId(note.getTitle(), false));
       note.setUrl("");
+      note.setAuthor(currentUser);
```
Should be centralized in Service layer, since the identity ACL is passed to Service (remember, in REST, no business logic should be done, which unfortunately not the case of a lot of stuff in this REST endpoint, but at least we need to avoid making it worse)

```suggestion
```

boubaker commented on notes-webapp/src/main/webapp/vue-app/notes/components/NoteContentTableItem.vue:
```diff
@@ -0,0 +1,24 @@
+<template>
```
Missing license header

boubaker commented:
Ok for me, thanks. Can you please squash the new commits added by PR review and add a significant title for each chunk of modifications.

use commit message of type :
`fix: ..... - Meeds-io/MIPs#71`
